Support for meta txs

oed opened the pull request:
This is my first go at adding meta tx support. It provides two new signers: `TxRelaySigner` and `MetaIdentitySigner`. These signers are different from the rest of the signers in that they have a function called `metaSignRawTx` instead of `signRawTx`, which is standard in the other signers. This is because we need additional data and that the transaction will not be signed in the same way as before.

### Usage
What the user does (uport-mobile):
```js
var proxyAddress = // the address of your proxy contract
var metaIdentityManagerAddress = // the address of the metaIdentityManager contract
var relayAddress = // the address of the txRelay contract
var txSenderAddress = // the address of the service that is sending your tx i.e. sensui
var keyPair = // your keypair that will be used to meta sign a transaction

var relaySigner = new TxRelaySigner(keypair, relayAddress, txSenderAddress);
var signer = new IdentityManagerSigner(proxyAddress, relaySigner, metaIdentityManagerAddress);

var rawTx = // a raw tx that you want to send
// IMPORTANT - the raw tx above has to have the nonce coresponding to your 
// keypairs address in the txRelay contract. If different the tx will fail.

signer.signRawTx(rawTx, (error, metaSignedRawTx) => {
  // send tx to sensui
})
```
The raw tx that is returned can then be sent to sensui which can sign the tx with a simpleSigner:
```js
var metaSignedRawTx = // the meta signed tx from above
// IMPORTANT - since the user won't know the nonce of the sender address it will have to provide it
var Transaction = require('ethereumjs-tx');
var tx = new Transaction(Buffer.from(metaSignedRawTx, 'hex'))
tx.nonce = // the nonce of sender address
var rawTx = tx.serialize().toString('hex')

var keypair = // sender keypair
var signer = SimpleSigner(keypair)
signer.signRawTx(rawTx, (error, signedRawTx) => {
  // send tx to ethereum network
})
```
#### Veryfy signature of a rawMetaTx
The relay service will probably want to verify that the metaTx it's relaying has a valid signature in order to avoid sending invalid txs. This can be done in the following way:
```js
var decodedMetaTx = TxRelaySigner.decodeMetaTx(metaSignedRawTx)
txRelay.getNonce(decodedMetaTx.claimedAddress).then(nonce => {
  var validMetaSig = TxRelaySigner.isMetaSignatureValid(txRelay.address, decodedMetaTx, nonce, simpleSigner.getAddress())
  if (validMetaSig) {
    // send the tx
  }
})
```
#### Use IdentityManager without metaTxs
In the MetaIdentityManager contract it is possible to sign transactions directly. If you want to sign a tx in this way you can do the following:
```js
var simpleSigner = SimpleSigner(keypair)
var signer = new IdentityManagerSigner(proxyAddress, simpleSigner, metaIdentityManagerAddress);

var rawTx = // a raw tx that you want to send
signer.signRawTx(rawTx, (error, metaSignedRawTx) => {
  // send tx to sensui
})
```
## Stuff left to do
I want to add integration tests that uses truffle to make sure that the signers work together with the contracts. This will prevent things such as the value bug.

christianlundkvist commented on lib/tx_relay_signer.js:
```diff
@@ -0,0 +1,64 @@
+
+  var nonce = txCopy.nonce.toString('hex');
+  var to = txCopy.to.toString('hex');
+  var data = txCopy.data.toString('hex');
```
No value here? We want to make sure we can send ETH from the proxy.

oed replied:
That would be handled by the metaIdentityManager, meaning that the value is in the `data` buffer.

christianlundkvist replied:
Hmm, if this is the case should we also have a helper function to package the users initial function in terms of `(destination, value, data)` into the data field of the above raw transaction?

christianlundkvist replied:
Oh, is this what the `MetaIdentitySigner` does? Need to look more closely here :)

christianlundkvist commented on lib/tx_relay_signer.js:
```diff
@@ -0,0 +1,64 @@
+    callback("Nonce, to, and data have to be specified in the tx")
+  } else {
+    // Tight packing, as Solidity sha3 does
+    hashInput = this.txRelayAddress + pad(nonce) + to + data + txSender.slice(2);
```
How is "pad defined?"

oed replied:
It's in the bottom of this file

oed commented:
@naterush You're right, updating the description :)

naterush commented on lib/meta_identity_signer.js:
```diff
@@ -0,0 +1,39 @@
+              "nonce": txCopy.nonce,
+              "to": this.metaIdentityManagerAddress
+              }
+  var rawForwardTx = txutils.functionTx(MetaIdentityManagerAbi, "forwardTo",
```
@oed does this only work with "forwardTo" as of now? Aka, would I be able to use meta-tx for the "addOwner" function in the metaIdentityManager, if I wanted to.

oed replied:
Good point! I think those functions can be called by only using the `TxRelaySigner` but need to take a closer look.

naterush commented on lib/meta_identity_signer.js:
```diff
@@ -0,0 +1,39 @@
+              "gasPrice": txCopy.gasPrice,
+              "gasLimit": txCopy.gasLimit,
+              "value": 0,
+              "nonce": txCopy.nonce,
```
Ok, just to clarify my understanding, this means that the nonce that must be provided here is the nonce of the keypairAddress in the txRelay contract, right? Thanks.

oed replied:
Exactly

naterush replied:
Ok, cool. This is a semi-unrelated question, but would this be hard to get? Aka, in the process of creating the transaction, would it be added difficulty to get this nonce?

oed replied:
This lib doesn't interact at all with the blockchain as it is now, and I don't think it should since that would impose a specific way of getting data from the blockchain.

naterush replied:
@oed how would they know their nonce then? Would uPort or some service give it to them?

oed replied:
That's up to the user of this lib to decide. This lib is only concerned with signing not any networking.

oed commented:
@christianlundkvist Right, the reason I have the gas price/limit there is because it also was in the proxy signer. But I don't think it's a problem really since sensui can change those values before it sends the tx.

pelle commented:
@oed Great work. Just a few questions and a suggestion for cleaning up the api.

What is the difference between `TxRelaySigner` and `MetaIdentitySigner`? It's not obvious to me, without fully reading through the code.

The primary ask besides that is to move `txSenderAddress` to the constructor and thus turn it into an instance variable. This allows you to have the exact same signature for this as for the others making it composable. I don't see this ever changing for a particular instantiation either.

christianlundkvist commented on lib/meta_identity_signer.js:
```diff
@@ -0,0 +1,39 @@
+    [ util.addHexPrefix(this.signer.getAddress()),
+      util.addHexPrefix(this.proxyAddress),
+      util.addHexPrefix(finalDestination.toString('hex')),
+      util.bufferToInt(txCopy.value),
```
Ah here is the value! Didn't see it because Github was collapsing this file :)

oed commented:
@pelle The `TxRelaySigner` creates the meta signature and puts that into a transaction along with the data that is passed. In the future there might be other contracts that use the TxRelay contract for supporting metaTxs. `MetaIdentitySigner` simply creates the wrapper transaction that is needed for the MetaIdentityManager, and is very similar to what ProxySigner does.

Moving `txSenderAddress` makes a lot of sense, thanks!

oed commented:
Alright, I moved `txSenderAddress` to the constructor of `TxRelaySigner` and renamed the signing functions to `signRawTx` as all other signers. I also renamed `MetaIdentitySigner` to `IdentityManagerSigner`. This is pretty nice because now you can use a `SimpleSigner` in `IdentityManagerSigner` if you don't want to use the metaTx functionallity.

I also updated the post on top to reflect this.

naterush commented:
@oed checked out the integration tests. They look good - except I wonder if we should test to confirm some "failing" stuff here (e.g. wrong person tries to sign and/or forward, etc), or is it sufficient to do so w/in the IdentityManager tests itself.

oed commented:
@naterush yeah I think we only want to check that the signing code works. Testing for errors in the contracts isn't really needed here.

naterush commented on test/integration/identity_manager.js:
```diff
@@ -0,0 +1,197 @@
+      })
+      var rawTx = txutils.functionTx(testRegArtifact.abi, 'register', [testNum], wrapperTx)
+      var metaSignedRawTx = await identityManagerSigner.signRawTxAsync(rawTx)
+      var signedRawTx = await simpleSigner.signRawTxAsync(metaSignedRawTx)
```
Just to clarify my understanding, the data being signed twice here is 1. by the user and then 2. by the relayer of the transaction, right?

oed replied:
Correct :)

oed replied:
Will add some comments to make this more clear

oed commented:
We need a way of verifying signature of the meta tx. Otherwise sensui could be wasting gas sending invalid txs.

oed commented:
Added helper function to verify that a metaTx has a valid signature.

christianlundkvist commented on test/integration/identity_manager.js:
```diff
@@ -0,0 +1,214 @@
+      })
+      var rawTx = txutils.functionTx(testRegArtifact.abi, 'register', [testNum], wrapperTx)
+      // tx is meta signed on users device
+      var metaSignedRawTx = await identityManagerSigner.signRawTxAsync(rawTx)
```
I feel like we should be using txRelaySigner rather than `identityManagerSigner` here. AFAICT `decodeMetaTx` doesn't make sense for a transaction that is signed by `identityManagerSigner`.

christianlundkvist replied:
Yeah I feel like the rest of the test works if we use a `txRelaySigner` here.  Does this test pass?

oed replied:
The `identityManagerSigner` has been passed the `relaySigner` (look in the before block) and uses that internally to sign the tx. 
What the `identityManagerSigner` does is adding the wrapper call that is needed to send a tx through the IdentityManager, it then delegates the signing to it's internal signer that has been passed to it in the constructor.

christianlundkvist replied:
@oed Got it thanks!

oed commented on lib/identity_manager_signer.js:
```diff
@@ -0,0 +1,39 @@
+              "nonce": txCopy.nonce,
+              "to": this.metaIdentityManagerAddress
+              }
+  var rawForwardTx = txutils.functionTx(MetaIdentityManagerAbi, 'forwardTo',
```
Yeah, that is probably a good idea. The reason I haven't added it is that I'm not sure how the code in the mobile app would handle such txs. Another approach is to create a raw tx directly to the IdentityManager and then sign that only with the `TxRelaySigner`.

christianlundkvist replied:
Ok we can add that later if we need to.

oed commented on lib/identity_manager_signer.js:
```diff
@@ -0,0 +1,39 @@
+      util.addHexPrefix(finalDestination.toString('hex')),
+      util.bufferToInt(txCopy.value),
+      util.addHexPrefix(txCopy.data.toString('hex')) ], wrapperTx)
+  this.signer.signRawTx(rawForwardTx, callback);
```
`this.signer` can either be a `SimpleSigner` or a `TxRelaySigner` depending on what you pass into the constructor of this object. This allows us to easily switch between regular txs and metaTxs.

christianlundkvist commented on lib/tx_relay_signer.js:
```diff
@@ -0,0 +1,94 @@
+
+TxRelaySigner.isMetaSignatureValid = function(txRelayAddress, decodedMetaTx, nonce, sender) {
+  // make sure nonce is a string
+  nonce += '';
```
If it isn't a string we should explode here. If it's not a string it's likely a number, hence decimal and not hex, so it would mess up the rest of the stuff right?

oed replied:
The reason we want it as a string is because the `pad` function used on the next line only handles strings. Maybe there is some better way to do this though?

christianlundkvist replied:
@oed What I meant was that right now you are casting it to a string if it's not a string (i.e. if it's a number `123` it will be converted to a string `'123'`). This will cause errors.

It's better to check something like

```
if (typeof foo !== "string") {throw;} //or similar
```

pelle commented on lib/generators/key_pair.js:
```diff
@@ -18,7 +17,7 @@ function fromPrivateKey(privateKey) {
 }
 function generate (callback) {
-  Random.randomBytes(32, function(error, rand) {
+  Random.naclRandom(32, function(error, rand) {
```
@oed this should actually be Random.randomBytes as it can be configured to use a different generator such as the one used by RN.

oed replied:
This was a stupid change I did, should have asked before. But the reason I did it was because I got a very strange error when running the tests. 
Running all tests the `MetaIdentityManager` integration tests fail because `randomBytes` is undefined. However running just the failing test by itself works fine. 
I'm trying to debug it today but no luck yet.

pelle replied:
You can do `Random.randomBytes = Random.naclRandom`

oed replied:
That is done in random.js but are you saying that I should do it in key_pair.js?

pelle replied:
@oed It is probably being unset somewhere. I know it's not the best design, but it's designed to be configurable. So just try resetting it.

oed commented:
Some small things left to do before this can be merged:

- [ ] Change to use uport-identity from npm (now uses the metaTx PR from gh)
- [ ] Add usage from the first comment above to README
- [ ] Seems like the MetaIM or TxRelay is not working in the new PR, debuging

pelle commented on lib/generators/key_pair.js:
```diff
@@ -17,7 +17,10 @@ function fromPrivateKey(privateKey) {
 }
 function generate (callback) {
-  Random.naclRandom(32, function(error, rand) {
+  if (!Random.randomBytes) {
```
This is good for now. We should think of a better way of configuring alternate versions in the future. But don't worry for now.

christianlundkvist commented:
Ok it looks good!
